src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs

```diff
@@ -73,5 +85,54 @@ public async Task<IEnumerable<AuthFlowResult>> GetTokenAsync()
 
             return resultList;
         }
+
+        /// <summary>
+        /// Run the auth mode in a separate task and
+        /// poll to see if we hit global timeout before the auth flow completes.
+        /// </summary>
+        /// <param name="authFlow">Auth Flow that we are executing.</param>
+        /// <returns>AuthFlowResult containing Error if CLI times out; else actual result of the AuthFlow.</returns>
+        private async Task<AuthFlowResult> GetTokenAndPoll(IAuthFlow authFlow)
+        {
+            var flowResult = Task.Run(() => authFlow.GetTokenAsync());
+            var authFlowName = authFlow.GetType().Name;
+            var oneMinute = TimeSpan.FromMinutes(1);
+            while (!flowResult.IsCompleted)
+            {
+                if (GlobalTimeoutManager.GetElapsedTime() >= oneMinute)
+                {
+                    this.logger.LogWarning($"Waiting on {authFlowName} authentication." +
+                        $"Timeout in {GlobalTimeoutManager.GetRemainingTime():hh\\:mm\\:ss}");
+                }
+
+                if (GlobalTimeoutManager.HasTimedout())
+                {
+                    GlobalTimeoutManager.StopTimer();
+                    this.logger.LogWarning("AzureAuth has timed out!");
+                    AuthFlowResult timeoutResult = new AuthFlowResult(null, null, authFlow.GetType().Name);
+                    timeoutResult.Errors.Add(new TimeoutException($"The application has timed out while waiting on {authFlowName}"));
+                    /* Note that though the task running the auth flow will be killed once we return from this method,
+                     * the interactive auth prompt will be killed as we exit the application (possibly due to the way GC works).
+                     */

```

Comment:
Regarding static class - I agree with you that in most cases we should avoid. For this case though it made sense to me because of two reasons:
1. Having timeout related stuff in AuthFlowExecutor would violate the Single Responsibility Principle as then that class has to execute the auth flow and also do housekeeping work for timeout.
2. This is one of the instance where a class should be available globally because the timeout itself is 'global'. Static might be a good option that satisfies the need. The other way would be to have a normal class and pass instances and make it singleton in order to avoid being duplicated which is also fine - just a toss up between two approaches.  
Please let me know if Singleton approach looks better and I am missing something that could be very bad with this static class and we can revisit the design in that case.